File: controllers/userController.js
const bcrypt = require('bcryptjs');
const User = require('../models/userModel');
const dotenv = require('dotenv');

dotenv.config();    // load environment variables


/**
 * Register a new user.
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
exports.registerUser = async (req, res) => {
  const { username, email, password, adminCode } = req.body;
  try {    
    // without requiring an admin code, we can do a two-in-one and admin-ify the first 
    // user that's added.
    const numUsers = await User.countUsers();
    let isAdmin = 0;
    let isApproved = 0;

    // check if the user should be an admin
    if (adminCode && adminCode === process.env.ADMIN_CODE) {
      isAdmin = 1;
      isApproved = 1;
    } else if (numUsers === 0) {
        // first user is an admin
        isAdmin = 1;
        isApproved = 1;
      }
      else {
        // standard users are assumed anonymous (i.e a QR code at a local cafe)
        // and require approval for any sort of access.
        isApproved = 0;
      }

    // With the role sorted out, we can proceed with creating the user 
    const hashedPassword = await bcrypt.hash(password, 10);
    const user = await User.create(username, email, hashedPassword, isAdmin, isApproved);

    res.status(201).json({
      message: 'User successfully registered!',
      user: {
        id: user.id,
        username: user.username,
        email: user.email,
        isAdmin: user.isAdmin,
        isApproved: user.isApproved,
      },
    });
  } catch (err) {     // print an error here for easier stack trace readability
    console.error('Registration error:', err);
    if (err.code === 'SQLITE_CONSTRAINT') {
      res.status(400).json({ error: 'Username/email already in use' });
    } else {
      res.status(500).json({ error: 'Internal server error' });
    }
  }
};

/**
 * Get user profile information.
 *
 * @param {Object} req - Express request object.
 * @param {Object} res - Express response object.
 */
exports.getUserProfile = async (req, res) => {
  
  // const username = req.user;
  const username = req.user.username; // populated from authMiddleware and jwtService
  // try to find the username, error if we can't find them
  try {
    const user = await User.findByUsername(username);
    if(user) {
      res.json({ username: user.username, email: user.email, isAdmin: user.isAdmin });
    } else {
      res.status(404).json({error: 'User not found'});
    }
  }
  catch {
    console.error('Error fetching user profile:', err);
    res.status(500).json({ error: 'Internal server error!'});
  }
};

// TODO: delete user, update user profile, profile page

File: python/controllers/user_controller.py
"""
User Controller Module

Handles server behavior for requests and responses involved with
user registration and profile management.
"""
import os
from flask import Blueprint, request, jsonify
import bcrypt
from dotenv import load_dotenv
from models.user_model import UserModel

# Load environment variables
load_dotenv()

class UserController:
    """
    Controller for handling user-related operations.
    Uses a class-based approach for better organization and OOP principles.
    """
    
    def __init__(self, user_model=None):
        """
        Initialize with dependency injection for better testability.
        
        Args:
            user_model: The user model to use (defaults to UserModel if None)
        """
        self.user_model = user_model or UserModel()
        
    def register_user(self):
        """
        Register a new user.
        
        Returns:
            JSON response with user information or error
        """
        data = request.get_json()
        username = data.get('username')
        email = data.get('email')
        password = data.get('password')
        admin_code = data.get('adminCode')
        
        try:
            # Count users to determine if this is the first user
            num_users = self.user_model.count_users()
            is_admin = 0
            is_approved = 0
            
            # Check if the user should be an admin
            if admin_code and admin_code == os.environ.get('ADMIN_CODE'):
                is_admin = 1
                is_approved = 1
            elif num_users == 0:
                # First user is automatically an admin
                is_admin = 1
                is_approved = 1
            else:
                # Standard users require approval
                is_approved = 0
                
            # Hash the password
            hashed_password = bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
            
            # Create the user
            user = self.user_model.create(username, email, hashed_password, is_admin, is_approved)
            
            return jsonify({
                'message': 'User successfully registered!',
                'user': {
                    'id': user.id,
                    'username': user.username,
                    'email': user.email,
                    'is_admin': user.is_admin,
                    'is_approved': user.is_approved
                }
            }), 201
            
        except Exception as err:
            # Handle database constraint violations
            if hasattr(err, 'orig') and getattr(err.orig, 'sqlite_errorcode', None) == 19:  # SQLITE_CONSTRAINT
                return jsonify({'error': 'Username/email already in use'}), 400
            else:
                # Log the error for debugging
                print(f"Registration error: {err}")
                return jsonify({'error': 'Internal server error'}), 500
                
    def get_user_profile(self):
        """
        Get user profile information using the username from the authenticated user.
        
        Returns:
            JSON response with user information or error
        """
        # In Flask, the user would be available from the auth middleware
        username = request.user.get('username')
        
        try:
            user = self.user_model.get_by_username(username)
            if user:
                return jsonify({
                    'username': user.username,
                    'email': user.email,
                    'is_admin': user.is_admin
                })
            else:
                return jsonify({'error': 'User not found'}), 404
        except Exception as err:
            print(f"Error fetching user profile: {err}")
            return jsonify({'error': 'Internal server error!'}), 500

# Create a Flask blueprint for user routes
user_bp = Blueprint('user', __name__)

# Create controller instance
user_controller = UserController()

# Define routes
@user_bp.route('/register', methods=['POST'])
def register():
    """Route for user registration"""
    return user_controller.register_user()

@user_bp.route('/profile', methods=['GET'])
def profile():
    """Route for getting user profile, requires authentication"""
    # The auth middleware will need to be applied to this route
    return user_controller.get_user_profile()